compiler/src/linking/link.re

```diff
@@ -0,0 +1,529 @@
+  | GlobalSet =>
+    let internal_name = Expression.global_set_get_name(expr);
+    Expression.global_set_set_name(
+      expr,
+      Hashtbl.find(local_names, internal_name),
+    );
+    globalize_names(local_names, Expression.global_set_get_value(expr));
+  | Load => globalize_names(local_names, Expression.load_get_ptr(expr))
+  | Store =>
+    globalize_names(local_names, Expression.store_get_ptr(expr));
+    globalize_names(local_names, Expression.store_get_value(expr));
+  | AtomicRMW
+  | AtomicCmpxchg
+  | AtomicWait
+  | AtomicNotify
+  | AtomicFence
+  | SIMDExtract
+  | SIMDReplace
+  | SIMDShuffle
+  | SIMDTernary
+  | SIMDShift
+  | SIMDLoad
+  | SIMDLoadStoreLane
+  | SIMDWiden
+  | Prefetch
+  | MemoryInit
+  | DataDrop => failwith("NYI")
+  | MemoryCopy =>
+    globalize_names(local_names, Expression.memory_copy_get_dest(expr));
+    globalize_names(local_names, Expression.memory_copy_get_source(expr));
+    globalize_names(local_names, Expression.memory_copy_get_size(expr));
+  | MemoryFill =>
+    globalize_names(local_names, Expression.memory_fill_get_dest(expr));
+    globalize_names(local_names, Expression.memory_fill_get_value(expr));
+    globalize_names(local_names, Expression.memory_fill_get_size(expr));
+  | Const => ()
+  | Unary => globalize_names(local_names, Expression.unary_get_value(expr))
+  | Binary =>
+    globalize_names(local_names, Expression.binary_get_left(expr));
+    globalize_names(local_names, Expression.binary_get_right(expr));
+  | Select =>
+    globalize_names(local_names, Expression.select_get_if_true(expr));
+    globalize_names(local_names, Expression.select_get_if_false(expr));
+    globalize_names(local_names, Expression.select_get_condition(expr));
+  | Drop => globalize_names(local_names, Expression.drop_get_value(expr))
+  | Return => globalize_names(local_names, Expression.return_get_value(expr))
+  | MemorySize => ()
+  | MemoryGrow =>
+    globalize_names(local_names, Expression.memory_grow_get_delta(expr))
+  | Unreachable => ()
+  | Pop
+  | RefNull
+  | RefIs
+  | RefFunc
+  | RefEq
+  | Try
+  | Throw
+  | Rethrow => failwith("NYI")
+  | TupleMake =>
+    let num_operands = Expression.tuple_make_get_num_operands(expr);
+    for (i in 0 to num_operands - 1) {
+      globalize_names(
+        local_names,
+        Expression.tuple_make_get_operand_at(expr, i),
+      );
+    };
+  | TupleExtract =>
+    globalize_names(local_names, Expression.tuple_extract_get_tuple(expr))
+  | I31New
+  | I31Get
+  | CallRef
+  | RefTest
+  | RefCast
+  | BrOn
+  | RttCanon
+  | RttSub
+  | StructNew
+  | StructGet
+  | StructSet
+  | ArrayNew
+  | ArrayGet
+  | ArraySet
+  | ArrayLen
+  | RefAs => failwith("NYI")
+  };
+};
+
+let table_offset = ref(0);
+let module_id = ref(0);
+
+let link_all = (linked_mod, dependencies) => {
+  table_offset := 0;
+
+  let link_one = dep => {
+    let local_names: Hashtbl.t(string, string) = Hashtbl.create(128);
+
+    let wasm_mod = Hashtbl.find(modules, dep);
```

Comment:
Question: Does `.find()` raise if the key is not found? If so, do we want to defensively code around it?

Reply:
It does. This couldn't ever fail though because the hash is built from the list that is being iterated over. We could add a check here to raise an error, though it'd be a pretty similar one (some form of `Not_found`).